confuse/core.py

```diff
@@ -295,6 +294,26 @@ def values(self):
 
     # List/sequence emulation.
 
+    def sequence(self):
+        """Iterates over the subviews contained in lists from the *first*
+        source at this view. If the object for this view in the first source
+        is not a list or tuple, then a `ConfigTypeError` is raised.
+        """
+        try:
+            collection, _ = self.first()
+        except NotFoundError:
+            return
+        if not isinstance(collection, (list, tuple)):
+            raise ConfigTypeError(
+                u'{0} must be a list, not {1}'.format(
+                    self.name, type(collection).__name__
+                )
+            )
+
+        # Yield all the indices in the sequence.
+        for index in range(len(collection)):
+            yield self[index]
```

Comment:
Maybe I'm missing something silly, but any particular reason why this wouldn't do the same thing (because we know we have a `list` or a `tuple`)?

```py
for value in collection:
    yield value
```

Reply:
In terms of `__iter__` raising the exception lazily, to get the immediate error, I think we would have to switch from using a generator function to defining `__iter__` and `__next__` separately, with `__iter__` doing the type check, raising an error if necessary, and otherwise returning `self`. Then we could define `__next__` to be like the current `__iter__`. I'm not sure if this would gain us much, though, and it might lead to duplicating the type checking code, unless we broke that out into separate methods like `is_dict_view` and `is_list_view`.

To address your comment above, changing to
```python
for value in collection:
    yield value
```
means that we are returning the value stored in the subview. However, the original `yield self[index]` returns the subview itself, and is how the iteration interface was defined. I had the same thought when I was moving the list view iteration code to its own method and seeing if I could simplify it, but yielding the value changes the output. Luckily the test code catches that. 😄

Reply:
Oops, you're right! Thanks for pointing all this out.